programs/registry/src/delegate/process_delegate.rs

```diff
@@ -0,0 +1,532 @@
+    }
+
+    #[test]
+    fn test_delegate_or_undelegate_undelegate_pass() {
+        let protocol_config = ProtocolConfig {
+            ..Default::default()
+        };
+
+        let mut forester_pda = get_test_forester_account();
+        let delegate_account = get_test_delegate_account_with_context();
+        let authority = delegate_account.delegate_account.owner;
+        let forester_pda_pubkey = delegate_account
+            .delegate_account
+            .delegate_forester_delegate_account
+            .unwrap();
+        let delegate_amount = 50;
+        let current_slot = 10;
+        let no_sync = true;
+
+        let result = delegate_or_undelegate::<false>(
+            &authority,
+            &protocol_config,
+            delegate_account,
+            &forester_pda_pubkey,
+            &mut forester_pda,
+            delegate_amount,
+            current_slot,
+            no_sync,
+        )
+        .unwrap();
+
+        let (input_delegate_pda, output_delegate_pda) = result;
+        assert_eq!(input_delegate_pda.compressed_account.owner, crate::ID);
+        assert_eq!(output_delegate_pda.compressed_account.owner, crate::ID);
+
+        let expected_delegate_account = DelegateAccount {
+            delegated_stake_weight: delegate_account.delegate_account.delegated_stake_weight
+                - delegate_amount,
+            pending_undelegated_stake_weight: delegate_account
+                .delegate_account
+                .pending_undelegated_stake_weight
+                + delegate_amount,
+            pending_epoch: protocol_config.get_current_epoch(current_slot),
+            ..delegate_account.delegate_account
+        };
+
+        let deserialized_delegate_account = DelegateAccount::deserialize(
+            &mut &output_delegate_pda
+                .compressed_account
+                .data
+                .as_ref()
+                .unwrap()
+                .data[..],
+        )
+        .unwrap();
+        assert_eq!(deserialized_delegate_account, expected_delegate_account);
+    }
+
+    #[test]
+    fn test_delegate_or_undelegate_undelegate_fail() {
+        let authority = Pubkey::new_unique();
+        let protocol_config = ProtocolConfig {
+            ..Default::default()
+        };
+        let forester_pda_pubkey = Pubkey::new_unique();
+        let mut forester_pda = get_test_forester_account();
+        let delegate_account = get_test_delegate_account_with_context();
+        let delegate_amount = u64::MAX;
+        let current_slot = 10;
+        let no_sync = true;
+
+        let result = delegate_or_undelegate::<false>(
+            &authority,
+            &protocol_config,
+            delegate_account,
+            &forester_pda_pubkey,
+            &mut forester_pda,
+            delegate_amount,
+            current_slot,
+            no_sync,
+        );
+
+        assert!(matches!(result, Err(error) if error == RegistryError::InvalidAuthority.into()));
```

Comment:
BTW, this should also work

```suggestion
        assert!(matches!(result, Err(RegistryError::InvalidAuthority)));
```

(no need to change, just saying)